File: rust/src/code/combination_sum.rs
pub struct Solution {}

impl Solution {
    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = Vec::new();
        Solution::bfs(0, Vec::<i32>::new(), 0, target, &candidates, &mut res);

        return res;
    }

    fn bfs(
        pos: usize,
        mut subset: Vec<i32>,
        accu: i32,
        target: i32,
        candidates: &Vec<i32>,
        res: &mut Vec<Vec<i32>>,
    ) {
        if accu == target {
            res.push(subset.clone());
            return;
        } else if accu > target || pos > candidates.len() - 1 {
            return;
        }

        for i in pos..candidates.len() {
            subset.push(candidates[i]);
            Self::bfs(
                i,
                subset.clone(),
                accu + candidates[i],
                target,
                candidates,
                res,
            );
            subset.pop();
        }
    }
}

#[cfg(test)]
pub mod test {
    use super::Solution;

    #[test]
    fn combination_sum_case_1() {
        // arragne
        let candidates = vec![2, 3, 5];
        let target = 8;

        // act
        let actual = Solution::combination_sum(candidates, target);
        let expected = vec![vec![2, 2, 2, 2], vec![2, 3, 3], vec![3, 5]];

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn combination_sum_case_2() {
        // arrange
        let candidates = vec![2];
        let target = 1;

        // act
        let actual = Solution::combination_sum(candidates, target);
        let expected: Vec<Vec<i32>> = Vec::new();

        // assert
        assert_eq!(expected, actual);
    }
}

File: python/combination_sum.py
class Solution:
    def combinationSum(self, candidates: list[int], target: int) -> list[list[int]]:
        """time complexity: O(2^(t/m)) where t is the target and m is the min value in candidates"""
        result: list[list[int]] = []
        subset: list[int] = []
        accu: int = 0
        self.dfs(0, accu, subset, result, candidates, target)
        return result

    def dfs(
        self,
        i: int,
        accu: int,
        subset: list[int],
        result: list[list[int]],
        candidates: list[int],
        target: int,
    ) -> None:
        if accu == target:
            result.append(subset.copy())
            return
        if i >= len(candidates) or accu > target:
            return
        subset.append(candidates[i])
        self.dfs(i, accu + candidates[i], subset, result, candidates, target)
        subset.pop()
        self.dfs(i + 1, accu, subset, result, candidates, target)


def test_two_sum_case_1():
    # arrange
    candidates: list[int] = [2, 3, 6, 7]
    target: int = 7
    expected: list[list[int]] = [[2, 2, 3], [7]]

    # act
    solution = Solution()
    actual = solution.combinationSum(candidates, target)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual


def test_two_sum_case_2():
    # arrange
    candidates: list[int] = [2, 3, 5]
    target: int = 8
    expected: list[list[int]] = [[2, 2, 2, 2], [2, 3, 3], [3, 5]]

    # act
    solution = Solution()
    actual = solution.combinationSum(candidates, target)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual


def test_two_sum_case_3():
    # arrange
    candidates: list[int] = [2]
    target: int = 1
    expected: list[list[int]] = []

    # act
    solution = Solution()
    actual = solution.combinationSum(candidates, target)

    # assert
    expected.sort()
    actual.sort()
    assert expected == actual
